### PR TITLE
Add https over http tunnel

### PR DESCRIPTION
**Pull request by mitchellalderson**

### CODE REVIEW

**Comment by mitchellalderson** [comment]
@kumarrishav the issue is, we have in place a safeguard against https over http, return null if `request protocol == https && proxy protocol == http`
should this be removed:
https://github.com/farawaysouthwest/axios-proxy-builder/blob/754a63a7bdb6ff8f801eabbaefd8b0bde1af2a90/src/utils.ts#L60

**Comment by kumarrishav** [comment]
not sure i got you here. 

if url is https over http proxy, then agent / tunnel need to be used. right?

File: src/proxy-builder.ts [hunk 1]
```diff
@@ -10,17 +12,32 @@ export interface RequestProxy {
...
+      proxy: {
+        host: hostname,
+        port: parseInt(port),
+      },
+    });
+
+    return {
+      httpsAgent: agent,
+    };
+  }
```
Comment (kumarrishav) [inline src/proxy-builder.ts]: seems like issue after fixing the merge conflict

Comment (mitchellalderson) [reply src/proxy-builder.ts]: No, issue isn't here, it's in our null checks in `./utils`. Its working as designed, just not for tunneling.

File: src/proxy-builder.ts [hunk 2]
```diff
@@ -10,17 +12,32 @@ export interface RequestProxy {
...
+  if (requestURLObject.protocol === "https:" && protocol === "http:") {
+    const agent = httpsOverHttp({
+      proxy: {
+        host: hostname,
+        port: parseInt(port),
+      },
+    });
+
+    return {
+      httpsAgent: agent,
```
Comment (kumarrishav) [inline src/proxy-builder.ts:41]: you need to add 
proxy: false
as well. i think

**Comment by mitchellalderson** [comment]
@kumarrishav pull down this branch and see if that works for you? I think this will resolve.

File: src/proxy-builder.ts (hunk 2 above)
Comment (mitchellalderson) [reply src/proxy-builder.ts:41]: hmm, let me look, I believe we just want it undefined to omit completely?

File: src/proxy-builder.test.ts [hunk 3]
```diff
@@ -12,22 +12,30 @@ describe("Test configureProxy", () => {
...
+        username: "",
+        password: "",
       },
-    } as RequestProxy);
+    });
+    expect(result.httpsAgent).toBeUndefined();
+  });
+
+  test("with tunnel agent", () => {
+    process.env.HTTP_PROXY = http_env;
```
Comment (kumarrishav) [inline src/proxy-builder.test.ts]: we need to test https over http proxy

process.env.HTTPS_PROXY = http_env;

File: src/proxy-builder.ts (hunk 2 above)
Comment (mitchellalderson) [reply src/proxy-builder.ts:41]: no, your right, it should be false! Axios's docs are not great!

File: src/proxy-builder.test.ts (hunk 3 above)
Comment (mitchellalderson) [reply src/proxy-builder.test.ts]: That seems like a bad practice? Wouldn't you just write your http proxy to http_proxy? We can of course write a case for this.

Comment (kumarrishav) [reply src/proxy-builder.test.ts]: that's what fix is for i.e if url is https, but underlying proxy is http.

ideally you need to add fallback to http incase of https proxy
https://github.com/request/request/blob/3c0cddc7c8eb60b470e9519da85896ed7ee0081e/lib/getProxyFromURI.js#L69

but here there is no fallback
https://github.com/farawaysouthwest/axios-proxy-builder/blob/master/src/utils.ts#L61

hence, i had to set 
process.env.HTTPS_PROXY = http_env;

Comment (kumarrishav) [reply src/proxy-builder.test.ts]: if you add the fallback logic in this code.,
then process.env.HTTP_PROXY = http_env; this will do the job and test

Comment (kumarrishav) [reply src/proxy-builder.test.ts]: i have added here https://github.com/farawaysouthwest/axios-proxy-builder/pull/4/commits/6537175f0ec4014b72b858801b3d48a4ba7dc8b4

Comment (mitchellalderson) [reply src/proxy-builder.test.ts]: Ok, I believe everything is in place?

Comment (mitchellalderson) [reply src/proxy-builder.test.ts]: You're too fast for me man!😄

File: src/utils.ts [hunk 4]
```diff
@@ -54,11 +54,23 @@ export const getProxyEnv = (requestURL: URL): string => {
 
   // get proxy based on request url's protocol
   if (requestURL.protocol == "http:") {
-    return process.env.HTTP_PROXY || process.env.http_proxy || null;
+    return (
+      process.env.HTTP_PROXY ||
+      process.env.http_proxy ||
+      process.env.HTTPS_PROXY ||
+      process.env.https_proxy ||
+      null
```
Comment (kumarrishav) [inline src/utils.ts]: let's stick with request implementation https://github.com/request/request/blob/3c0cddc7c8eb60b470e9519da85896ed7ee0081e/lib/getProxyFromURI.js#L62. i,e no fallback on https for http url

given that it's a battle-tested module

Comment (mitchellalderson) [reply src/utils.ts]: I was thinking the same, stand by...

File: src/proxy-builder.test.ts [hunk 5]
```diff
@@ -12,22 +12,29 @@ describe("Test configureProxy", () => {
...
+        username: "",
+        password: "",
       },
-    } as RequestProxy);
+    });
+    expect(result.httpsAgent).toBeUndefined();
+  });
+
+  test("with tunnel agent - with http env in proxy", () => {
+    process.env.HTTPS_PROXY = http_env;
```
Comment (kumarrishav) [inline src/proxy-builder.test.ts:28]: as you have added the fallback now, you can use

process.env.HTTP_PROXY = http_env;

Comment (kumarrishav) [reply src/proxy-builder.test.ts:28]: oh. seems like you removed the fallback. okay. we need to keep as it is

File: src/proxy-builder.test.ts (hunk 3 above)
Comment (kumarrishav) [reply src/proxy-builder.test.ts]: haha. closed that PR. merge conflict. messed up few things

File: src/proxy-builder.test.ts (hunk 5 above)
Comment (mitchellalderson) [reply src/proxy-builder.test.ts:28]: I don't believe so, with the fallback reverted to the way `request` did it? It fails tests.

**Review by kumarrishav — APPROVED** [review]

File: src/proxy-builder.test.ts (hunk 5 above)
Comment (kumarrishav) [reply src/proxy-builder.test.ts:28]: that's weird

Comment (kumarrishav) [reply src/proxy-builder.test.ts:28]: if you have fallback in place, then process.env.HTTP_PROXY = http_env shouldn't cause the issue 

anyway, let's go with this for now. we can cover it in another PR. i can give a try

**Pull request merged**